File: backend/src/Application/Users/Dtos/UserDetailsResponse.cs
using Application.Common;

using Domain.Entities;

namespace Application.Users.Dtos;

public sealed record UserDetailsResponse : BaseResponse, IResponse<User, UserDetailsResponse>
{
    public required string FirstName { get; init; }
    public required string LastName { get; init; }
    public required string Email { get; init; }
    public DateTimeOffset? LastLogin { get; init; }

    public IEnumerable<UserPermissionResponse> Permissions { get; private init; } = [];

    public static UserDetailsResponse FromEntity(User entity)
    {
        return new UserDetailsResponse
        {
            Id = entity.Id,
            FirstName = entity.FirstName,
            LastName = entity.LastName,
            Email = entity.Email,
            LastLogin = entity.LastLogin,
            Permissions = entity.UserPermissions.Select(UserPermissionResponse.FromEntity),
            CreatedAtUtc = entity.CreatedAtUtc
        };
    }
}

File: web/src/app/core/models/users/user-details-response.ts
export interface UserPermissionResponse {
  permission: string;
}

export interface UserDetailsResponse {
  id: number;
  firstName: string;
  lastName: string;
  email: string;
  lastLoginUtc: string | null;
  createdAtUtc: string;
  permissions: UserPermissionResponse[];
}
